draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+
+    def process(self):
+        """Make a mock catalog based on input PDF.
+
+        Returns
+        ----------
+        mock_catalog : :class:`containers.SpectroscopicCatalog`
+            Simulated catalog.
+        """
+
+        if self.rank == 0:
+            # Only rank zero is relevant.
+            # The number of sources in each redshift bin follows a multinomial
+            # distribution (reshape from (1,nz) to (nz) to make a 1D array):
+            global_source_numbers = np.random.multinomial(
+                self.nsources, self.global_z_weights
+            )
+        else:
+            # All processes must have a value for source_numbers:
+            global_source_numbers = None
+
+        # Send number of sources per redshift to local sections on each rank.
+        # Need to pass tuples. For some reason lists don't work.
+        # source_numbers has shape (self.ls)
+        source_numbers = np.zeros(self.ls, dtype=np.int)
+        self.comm_.Scatterv(
+            [
+                global_source_numbers,
+                tuple(self.ls_list),
+                tuple(self.lo_list),
+                MPI.DOUBLE,
+            ],
+            source_numbers,
+        )
+
+        # For each z bin in local section, draw a uniform random number in [0,1]
+        # for each source. This will determine which angular pixel the source
+        # is assigned to.
+        # Shape of rnbs: [self.ls=local # of z-bins][# of sources in each z-bin]
+        rnbs = [np.random.uniform(size=num) for num in source_numbers]
+
+        # For each source, determine index of pixel the source falls into.
+        # Shape: [self.ls=local # of z-bins][# of sources in each z-bin]
+        idxs = [np.digitize(rnbs[ii], self.cdf[ii]) for ii in range(len(rnbs))]
+
+        # If desired, generate random numbers to randomize position of sources
+        # in each z bin. These are uniform random numbers in [-0.5, 0.5], which
+        # will determine the source's relative displacement from the bin's
+        # mean redshift.
+        if not self.z_at_channel_centers:
+            rz = [np.random.uniform(size=num) - 0.5 for num in source_numbers]
+
+        # If desired, generate random numbers for z errors, as standard normals
+        # to be multiplied by appropriate standard deviation later
+        if (self.sigma_z is not None) or (self.sigma_z_over_1plusz is not None):
+            rzerr = [np.random.normal(size=num) for num in source_numbers]
+
+        # If desired, generate random numbers to randomize position of sources
+        # in each healpix pixel. These are uniform random numbers in [-0.5, 0.5],
+        # which will determine the source's relative displacement from the bin's
+        # central RA and dec.
+        if not self.srcs_at_pixel_centers:
+            rtheta = [np.random.uniform(size=num) - 0.5 for num in source_numbers]
+            rphi = [np.random.uniform(size=num) - 0.5 for num in source_numbers]
```

Comment:
I think these can all be moved within the loop over redshifts below, which avoids needing to store all the random shifts for all redshifts, and means they don't all need to be generated (i.e. you only need to generate them based on the conditional values).

Reply:
Good point, and same for the above comment - it was indeed possible to refactor all this to make it nicer